cmd/controller/sparsefilegenerator.go

```diff
@@ -0,0 +1,180 @@
+func GetSparseFileSize() int64 {
+
+	sparseFileSizeStr := os.Getenv(ENV_SPARSE_FILE_SIZE)
+	if len(sparseFileSizeStr) < 1 {
+		sparseFileSizeStr = SPARSE_FILE_DEFAULT_SIZE
+	}
+
+	sparseFileSize, econv := strconv.ParseInt(sparseFileSizeStr, 10, 64)
+	if econv != nil {
+		glog.Info("Error converting sparse file size:  ", sparseFileSizeStr)
+		return 0
+	}
+
+	return sparseFileSize
+}
+
+// InitializeSparseFiles will check if the sparse file exist or have to be
+//  created and will update or create the associated Disk CR accordingly
+func (c *Controller) InitializeSparseFiles() {
+	sparseFileDir := GetSparseFileDir()
+	sparseFileSize := GetSparseFileSize()
+	sparseFileCount := GetSparseFileCount()
+
+	if len(sparseFileDir) < 1 || sparseFileSize < 1 || sparseFileCount < 1 {
+		glog.Info("No sparse file path/size provided. Skip creating sparse files.")
+		return
+	}
+
+	for i := 0; i < sparseFileCount; i++ {
+		sparseFile := path.Join(sparseFileDir, fmt.Sprint(i)+"-"+SPARSE_FILE_NAME)
+		err := CheckAndCreateSparseFile(sparseFile, sparseFileSize)
+		if err != nil {
+			glog.Info("Error creating sparse file: ", sparseFile, "Error: ", err)
+			continue
+		}
+		c.MarkSparseDiskStateActive(sparseFile, sparseFileSize)
+	}
+}
+
+// CheckAndCreateSparseFile will reuse the existing sparse file if it already exists,
+//   for handling cases where NDM is upgraded or restarted. If the file doesn't exist
+//   a new file will be created.
+func CheckAndCreateSparseFile(sparseFile string, sparseFileSize int64) error {
+	sparseFileInfo, err := util.SparseFileInfo(sparseFile)
+	if err != nil {
+		err = util.SparseFileCreate(sparseFile, sparseFileSize)
+	} else {
+		glog.Info("Sparse file already exists: ", sparseFileInfo.Name())
+	}
+	return err
+}
+
+// GetSparseDiskUuid returns a fixed UUID for the sparse
+//  disk on a given node.
+func GetSparseDiskUuid(hostname string, sparseFile string) string {
+	return SPARSE_DISK_PREFIX + util.Hash(hostname+sparseFile)
+}
+
+// MarkSparseDiskStateActive will either create a Disk CR if it doesn't exist, or it will
+//   update the state of the existing CR as Active.  Note that, when the NDM is going being
+//   gracefully shutdown, all its Disk CRs are marked with State as Unknown.
+func (c *Controller) MarkSparseDiskStateActive(sparseFile string, sparseFileSize int64) {
+	// Fill in the details of the sparse disk
+	diskDetails := NewDiskInfo()
+	diskDetails.Uuid = GetSparseDiskUuid(c.HostName, sparseFile)
+	diskDetails.HostName = c.HostName
+
+	diskDetails.DiskType = SPARSE_DISKTYPE
+	diskDetails.Path = sparseFile
+
+	sparseFileInfo, err := util.SparseFileInfo(sparseFile)
+	if err != nil {
+		glog.Info("Error fetching the size of sparse file: ", err)
+		glog.Error("Failed to create a disk CR for sparse file: ", sparseFile)
+		return
+	}
+
+	diskDetails.Capacity = uint64(sparseFileInfo.Size())
+
+	//If a Disk CR already exits, update it. If not create a new one.
+	c.CreateDisk(diskDetails.ToDisk())
```

Comment:
Does this call always return success ?

Reply:
Using an existing method here. If there is an error creating the CR, it logs an error.

Reply:
The log in the next line may become ambiguous if the CreateDisk fails.